File: rust/src/spatial.rs
use std::collections::HashMap;

/// 均匀网格宽相位：统计半径范围内的候选实体对。
pub fn nearby(xs: &[f64], ys: &[f64], radius: f64, cell_size: f64) -> usize {
    if xs.len() != ys.len() || cell_size <= 0.0 || radius < 0.0 { return 0; }
    let mut grid: HashMap<(i64, i64), Vec<usize>> = HashMap::new();
    for i in 0..xs.len() {
        let cell = ((xs[i] / cell_size).floor() as i64, (ys[i] / cell_size).floor() as i64);
        grid.entry(cell).or_default().push(i);
    }
    let reach = (radius / cell_size).ceil() as i64;
    let r2 = radius * radius;
    let mut pairs = 0;
    for (&(cx, cy), indices) in &grid {
        for dx in -reach..=reach {
            for dy in -reach..=reach {
                if let Some(other) = grid.get(&(cx + dx, cy + dy)) {
                    for &i in indices {
                        for &j in other {
                            if j > i {
                                let ax = xs[i] - xs[j];
                                let ay = ys[i] - ys[j];
                                if ax * ax + ay * ay <= r2 { pairs += 1; }
                            }
                        }
                    }
                }
            }
        }
    }
    pairs
}

File: src/spatial.py
"""用于查找附近实体的简单接口。"""
from array import array
from edg_hot import _LIB, _ptr


def nearby(xs, ys, radius, cell_size=None):
    x, y = array('d', map(float, xs)), array('d', map(float, ys))
    if len(x) != len(y):
        raise ValueError('arrays must have same size')
    if radius < 0:
        raise ValueError('radius must be non-negative')
    cell = radius if cell_size is None else cell_size
    if cell <= 0:
        raise ValueError('cell size must be positive')
    if _LIB:
        return _LIB.edg_nearby(_ptr(x), _ptr(y), len(x), radius, cell)
    grid = {}
    for i, (px, py) in enumerate(zip(x, y)):
        key = (int(px // cell), int(py // cell))
        grid.setdefault(key, []).append(i)
    reach = int(radius / cell + 0.999999)
    r2, total = radius * radius, 0
    for (cx, cy), items in grid.items():
        for dx in range(-reach, reach + 1):
            for dy in range(-reach, reach + 1):
                for i in items:
                    for j in grid.get((cx + dx, cy + dy), ()):
                        if j > i and (x[i]-x[j])**2 + (y[i]-y[j])**2 <= r2:
                            total += 1
    return total